File: csharp/anagram/Anagram.cs
using System;
using System.Linq;

public class Anagram
{
    private readonly string baseWord;
    private readonly string orderedChars;

    public Anagram(string baseWord)
    {
        this.baseWord = baseWord;
        orderedChars = ToOrderedChars(baseWord);
    }

    public string[] FindAnagrams(string[] potentialMatches) =>
        potentialMatches
            .Where(IsDifferentWord)
            .Where(IsAnagram)
            .Distinct()
            .ToArray();

    private Func<string, bool> IsDifferentWord =>
        (word) => !AreEqualWords(word, baseWord);

    private Func<string, bool> IsAnagram =>
        (word) => AreEqualWords(ToOrderedChars(word), orderedChars);


    private bool AreEqualWords(string left, string right) =>
        left.Equals(right, StringComparison.InvariantCultureIgnoreCase);

    private string ToOrderedChars(string word) =>
        string.Concat(word.ToLowerInvariant().OrderBy(letter => letter));
}

File: fsharp/anagram/Anagram.fs
module Anagram

open System

let normalize (target : string) =
    let valueNormalized =
        target.ToLowerInvariant().ToCharArray()
        |> Array.sort

    new string(valueNormalized)

let unequal target other =
    let same =
        String.Equals(target, other, StringComparison.InvariantCultureIgnoreCase)
    
    not (same)
       
let findAnagrams sources target = 
    let isMatch source =
        (source |> normalize) = (target |> normalize) && target |> unequal source
    
    sources
    |> List.filter isMatch